### PR TITLE
multiple compilation units in Rust

### PR DESCRIPTION
**Pull request by etsal**
Add code to scx_utils that allows multiple compilation units for Rust schedulers.  Also refactor scx_layered to use the new API.

### CODE REVIEW

File: scheds/rust/scx_layered/src/bpf/util.bpf.c [hunk 1]
```diff
@@ -1,19 +1,23 @@
...
+#include <errno.h>
 #include <stdbool.h>
 #include <string.h>
 #include <bpf/bpf_core_read.h>
 #include <bpf/bpf_helpers.h>
 #include <bpf/bpf_tracing.h>
-#endif
+
+#include "intf.h"
+#include "timer.bpf.h"
```
Comment (hodgesds) [inline scheds/rust/scx_layered/src/bpf/util.bpf.c:19]: Why does util need to include `timer.bpf.h`, there's no timer being used at all IIRC.

File: scheds/rust/scx_layered/src/bpf/main.bpf.c [hunk 2]
```diff
@@ -9,16 +9,17 @@
...
-#include "timer.bpf.h"
-
 #include <errno.h>
 #include <stdbool.h>
 #include <string.h>
 #include <bpf/bpf_core_read.h>
 #include <bpf/bpf_helpers.h>
 #include <bpf/bpf_tracing.h>
 
+#include "intf.h"
```
Comment (hodgesds) [inline scheds/rust/scx_layered/src/bpf/main.bpf.c:19]: This looks so much better now, no random includes in the middle of the file!

**Review by hodgesds — APPROVED** [review]
LGTM

**Review by htejun — APPROVED** [review]

**Review by likewhatevs — APPROVED** [review]
this or an equivalent (or workaround) needs to work w/ all the build systems we use, but I like this. 

it'll make LSP work nicer also and nudge things towards being more c and less bpf c.

**Comment by sirlucjan** [comment]
```FAILED: Cargo.toml.__PHONY__ 
/usr/bin/meson --internal exe --unpickle /home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/meson-private/meson_exe_cargo_59c58fbc71558b247ce3e9f35a34239e102035cc.dat
while executing ['/usr/bin/cargo', 'build', '--manifest-path=../Cargo.toml', '--target-dir=.', '--quiet', '--release']
--- stdout ---

--- stderr ---
error: failed to run custom build command for `scx_layered v1.0.8 (/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/scheds/rust/scx_layered)`

Caused by:
  process didn't exit successfully: `/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/./release/build/scx_layered-c80ab1fab0b0a9ec/build-script-build` (exit status: 101)
  --- stdout
  scx_utils:clang=ClangInfo { clang: "/usr/bin/clang", ver: "18.1.8", arch: "x86_64" } ["-g", "-O2", "-Wall", "-Wno-compare-distinct-pointer-types", "-D__TARGET_ARCH_x86", "-mcpu=v3", "-mlittle-endian", "-idirafter", "/usr/lib/clang/18/include", "-idirafter", "/usr/local/include", "-idirafter", "/usr/include", "-I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/libbpf/src/usr/include", "-I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/libbpf/include/uapi", "-I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/./release/build/scx_layered-d12b2e7aba4f5a4f/out/scx_utils-bpf_h/arch/x86", "-I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/./release/build/scx_layered-d12b2e7aba4f5a4f/out/scx_utils-bpf_h", "-I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/./release/build/scx_layered-d12b2e7aba4f5a4f/out/scx_utils-bpf_h/bpf-compat"]
  /usr/bin/clang is version 18.1.8
  Building src/bpf/main.bpf.c
  /usr/bin/clang is version 18.1.8
  Building src/bpf/timer.bpf.c

  --- stderr
  thread 'main' panicked at scheds/rust/scx_layered/build.rs:14:10:
  called `Result::unwrap()` on an `Err` value: failed to build `src/bpf/timer.bpf.c`

  Caused by:
      0: Failed to compile /home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/./release/build/scx_layered-d12b2e7aba4f5a4f/out/bpf from src/bpf/timer.bpf.c
      1: Command `/usr/bin/clang -g -O2 -Wall -Wno-compare-distinct-pointer-types -D__TARGET_ARCH_x86 -mcpu=v3 -mlittle-endian -idirafter /usr/lib/clang/18/include -idirafter /usr/local/include -idirafter /usr/include -I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/libbpf/src/usr/include -I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/libbpf/include/uapi -I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/./release/build/scx_layered-d12b2e7aba4f5a4f/out/scx_utils-bpf_h/arch/x86 -I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/./release/build/scx_layered-d12b2e7aba4f5a4f/out/scx_utils-bpf_h -I/home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/./release/build/scx_layered-d12b2e7aba4f5a4f/out/scx_utils-bpf_h/bpf-compat -I /tmp/.tmpBOe8XC/bpf/src -fno-stack-protector -g -O2 -target bpf -c src/bpf/timer.bpf.c -o /home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/./release/build/scx_layered-d12b2e7aba4f5a4f/out/bpf` failed (exit status: 1)
      2: src/bpf/timer.bpf.c:27:22: error: use of undeclared identifier 'MAX_TIMERS'; did you mean 'MAX_LAYERS'?
            27 |         __uint(max_entries, MAX_TIMERS);
               |                             ^~~~~~~~~~
               |                             MAX_LAYERS
         /home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/libbpf/src/usr/include/bpf/bpf_helpers.h:13:39: note: expanded from macro '__uint'
            13 | #define __uint(name, val) int (*name)[val]
               |                                       ^
         src/bpf/intf.h:37:2: note: 'MAX_LAYERS' declared here
            37 |         MAX_LAYERS              = 16,
               |         ^
         src/bpf/timer.bpf.c:35:39: error: use of undeclared identifier 'MAX_TIMERS'; did you mean 'MAX_LAYERS'?
            35 |         if (timerw->key < 0 || timerw->key > MAX_TIMERS) {
               |                                              ^~~~~~~~~~
               |                                              MAX_LAYERS
         src/bpf/intf.h:37:2: note: 'MAX_LAYERS' declared here
            37 |         MAX_LAYERS              = 16,
               |         ^
         src/bpf/timer.bpf.c:39:36: error: use of undeclared identifier 'layered_timers'; did you mean 'layered_timer_cb'?
            39 |         struct layered_timer *cb_timer = &layered_timers[timerw->key];
               |                                           ^~~~~~~~~~~~~~
               |                                           layered_timer_cb
         src/bpf/timer.bpf.c:33:12: note: 'layered_timer_cb' declared here
            33 | static int layered_timer_cb(void *map, int key, struct timer_wrapper *timerw)
               |            ^
         src/bpf/timer.bpf.c:39:36: error: subscript of pointer to function type 'int (void *, int, struct timer_wrapper *)'
            39 |         struct layered_timer *cb_timer = &layered_timers[timerw->key];
               |                                           ^~~~~~~~~~~~~~
         src/bpf/timer.bpf.c:40:17: error: call to undeclared function 'run_timer_cb'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
            40 |         bool resched = run_timer_cb(timerw->key);
               |                        ^
         src/bpf/timer.bpf.c:42:39: error: incomplete definition of type 'struct layered_timer'
            42 |         if (!resched || !cb_timer || cb_timer->interval_ns == 0) {
               |                                      ~~~~~~~~^
         src/bpf/timer.bpf.c:39:9: note: forward declaration of 'struct layered_timer'
            39 |         struct layered_timer *cb_timer = &layered_timers[timerw->key];
               |                ^
         src/bpf/timer.bpf.c:48:12: error: incomplete definition of type 'struct layered_timer'
            48 |                         cb_timer->interval_ns,
               |                         ~~~~~~~~^
         src/bpf/timer.bpf.c:39:9: note: forward declaration of 'struct layered_timer'
            39 |         struct layered_timer *cb_timer = &layered_timers[timerw->key];
               |                ^
         src/bpf/timer.bpf.c:49:12: error: incomplete definition of type 'struct layered_timer'
            49 |                         cb_timer->start_flags);
               |                         ~~~~~~~~^
         src/bpf/timer.bpf.c:39:9: note: forward declaration of 'struct layered_timer'
            39 |         struct layered_timer *cb_timer = &layered_timers[timerw->key];
               |                ^
         src/bpf/timer.bpf.c:59:23: error: use of undeclared identifier 'MAX_TIMERS'; did you mean 'MAX_LAYERS'?
            59 |         bpf_for(timer_id, 0, MAX_TIMERS) {
               |                              ^~~~~~~~~~
               |                              MAX_LAYERS
         /home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/libbpf/src/usr/include/bpf/bpf_helpers.h:390:40: note: expanded from macro 'bpf_for'
           390 |                                 bpf_iter_num_new(&___it, (start), (end)),                       \
               |                                                                    ^
         src/bpf/intf.h:37:2: note: 'MAX_LAYERS' declared here
            37 |         MAX_LAYERS              = 16,
               |         ^
         src/bpf/timer.bpf.c:59:23: error: use of undeclared identifier 'MAX_TIMERS'; did you mean 'MAX_LAYERS'?
            59 |         bpf_for(timer_id, 0, MAX_TIMERS) {
               |                              ^~~~~~~~~~
               |                              MAX_LAYERS
         /home/lucjan/Pracownia/Repo/scx-scheds/scx-scheds-git/src/scx/build/libbpf/src/usr/include/bpf/bpf_helpers.h:398:51: note: expanded from macro 'bpf_for'
           398 |                 (___t && ((i) = *___t, (i) >= (start) && (i) < (end)));                         \
               |                                                                 ^
         src/bpf/intf.h:37:2: note: 'MAX_LAYERS' declared here
            37 |         MAX_LAYERS              = 16,
               |         ^
         src/bpf/timer.bpf.c:65:34: error: use of undeclared identifier 'MAX_TIMERS'; did you mean 'MAX_LAYERS'?
            65 |                 if (timer_id < 0 || timer_id > MAX_TIMERS) {
               |                                                ^~~~~~~~~~
               |                                                MAX_LAYERS
         src/bpf/intf.h:37:2: note: 'MAX_LAYERS' declared here
            37 |         MAX_LAYERS              = 16,
               |         ^
         src/bpf/timer.bpf.c:70:38: error: use of undeclared identifier 'layered_timers'; did you mean 'layered_timer_cb'?
            70 |                 struct layered_timer *new_timer = &layered_timers[timer_id];
               |                                                    ^~~~~~~~~~~~~~
               |                                                    layered_timer_cb
         src/bpf/timer.bpf.c:33:12: note: 'layered_timer_cb' declared here
            33 | static int layered_timer_cb(void *map, int key, struct timer_wrapper *timerw)
               |            ^
         src/bpf/timer.bpf.c:70:38: error: subscript of pointer to function type 'int (void *, int, struct timer_wrapper *)'
            70 |                 struct layered_timer *new_timer = &layered_timers[timer_id];
               |                                                    ^~~~~~~~~~~~~~
         src/bpf/timer.bpf.c:79:19: error: incomplete definition of type 'struct layered_timer'
            79 |                                      new_timer->init_flags);
               |                                      ~~~~~~~~~^
         src/bpf/timer.bpf.c:70:10: note: forward declaration of 'struct layered_timer'
            70 |                 struct layered_timer *new_timer = &layered_timers[timer_id];
               |                        ^
         src/bpf/timer.bpf.c:92:20: error: incomplete definition of type 'struct layered_timer'
            92 |                                       new_timer->interval_ns,
               |                                       ~~~~~~~~~^
         src/bpf/timer.bpf.c:70:10: note: forward declaration of 'struct layered_timer'
            70 |                 struct layered_timer *new_timer = &layered_timers[timer_id];
               |                        ^
         src/bpf/timer.bpf.c:93:20: error: incomplete definition of type 'struct layered_timer'
            93 |                                       new_timer->start_flags);
               |                                       ~~~~~~~~~^
         src/bpf/timer.bpf.c:70:10: note: forward declaration of 'struct layered_timer'
            70 |                 struct layered_timer *new_timer = &layered_timers[timer_id];
               |                        ^
         16 errors generated.
         
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

ninja: build stopped: subcommand failed.
```
@etsal Doesn't work properly.

**Comment by etsal** [comment]
@sirlucjan Yes, the build was failing because of a missing header.  Waiting for all tests to complete but ```scx_layered``` works fine. Also as a note, this change currently works with ```scx_layered```. The rest of the Rust schedulers use the pre-existing build code.

**Comment by sirlucjan** [comment]
Now everything is OK.

**Pull request merged**